Correct AMP Chartbeat Config

andrewscfc opened the pull request:
Resolves #4811 

**Overall change:** Removed some config supplied to the Chartbeat AMP element that is not specified in its documentation http://support.chartbeat.com/docs/integrations.html#ampimpl 

**Code changes:**

_Fix AMP Analytics_
- Removed `triggers` configuration as it was not specified in [chartbeat's documentation](http://support.chartbeat.com/docs/integrations.html#ampimpl )
- Updated tests to not expect the `trigger` configuration 

_Address Missing Dependency Eslint https://github.com/bbc/simorgh/issues/4811#issuecomment-559461053_
- Initially added `chartbeatConfig` to the dependencies to address the eslint error: bd1eb2e871be6a813fc822a6c09790eeb8089610
- This caused an infinite render loop: https://github.com/bbc/simorgh/pull/4861#discussion_r352550260 
- This was because `chartbeatConfig` is created on each render and react hooks' equality check in the hook dependencies will always compare a newly created object to a previous one, each with a different reference in memory (further reading: http://adripofjavascript.com/blog/drips/object-equality-in-javascript.html)
- ~~I have addressed this issue by [using`useMemo`](https://github.com/bbc/simorgh/pull/4861/files#diff-f85d75438a01e2cb63e2f407f289dfc6R20-R31) to ensure `chartbeatConfig` is only updated when the props change. This means its reference in memory will not change between each render.~~
  - ~~This was an approach recommended by a member of the react team to get around the need for a custom comparator where the object reference is being used for equality comparison: https://github.com/facebook/react/issues/14476#issuecomment-471199055~~
- Moved away from the above change ^ considering @jroebu14 's comment https://github.com/bbc/simorgh/pull/4861#issuecomment-560473946 and have reluctantly decided to disable the lint rule for [the one line](https://github.com/bbc/simorgh/pull/4861/files#diff-f85d75438a01e2cb63e2f407f289dfc6R37).
  - The issue the lint rule tries to catch is missing dependencies, we address this by extracting all the dependencies out to [one variable as an object](https://github.com/bbc/simorgh/pull/4861/files#diff-f85d75438a01e2cb63e2f407f289dfc6R20) and spreading it's values out as dependencies for [the `useEffect`](https://github.com/bbc/simorgh/pull/4861/files#diff-f85d75438a01e2cb63e2f407f289dfc6R37). This means that if we use additional props to [get the config](https://github.com/bbc/simorgh/pull/4861/files#diff-f85d75438a01e2cb63e2f407f289dfc6R31), we will have to modify the variable, automatically updating the dependencies.

---

- [x] I have assigned myself to this PR and the corresponding issues
- [x] Automated (jest and/or cypress) tests added (for new features) or updated (for existing features)
- [ ] If necessary, I have run the local E2E non-smoke tests relevant to my changes (`CYPRESS_APP_ENV=local CYPRESS_SMOKE=false npm run test:e2e:interactive`) 
Ran `npm run test:e2e` (smoke only) to verify tests no longer fail on errors from chartbeat amp: `[AmpAnalytics <unknown id>] Ignoring request for event. Request string not found: pageview`
- [x] This PR requires manual testing
There are no automated tests to check our integration with Chartbeat, we have decided on previous issues to cover this with manual testing.

andrewscfc committed:
WIP, remove undocumented config for
test http://support.chartbeat.com/docs/integrations.html#ampimpl

andrewscfc committed:
Update Chartbeat Amp tests to remove
unecessary config

andrewscfc committed:
Merge branch 'latest' into chartbeat-amp

andrewscfc commented:
Manual testing by me:
- Visited http://localhost.bbc.com:7080/news/articles/c0g992jmmkko.amp 
- Observed suitable ping requests with 'sensible looking' parameters:
<img width="456" alt="image" src="https://user-images.githubusercontent.com/9645462/69947046-bc271b80-14e4-11ea-8abe-e955b3105aef.png">

andrewscfc committed:
Add missing dependency for chartbeat
beacon dispatch

andrewscfc committed:
Merge branch 'chartbeat-amp' of github.com:bbc/simorgh into chartbeat-amp

andrewscfc committed:
Merge branch 'latest' into chartbeat-amp

rhenshaw56 commented on src/app/containers/ChartbeatAnalytics/index.jsx:
```diff
@@ -41,6 +41,7 @@ const ChartbeatAnalytics = ({ data }) => {
     service,
     sendCanonicalChartbeatBeacon,
     isCanonicalAndEnabled,
+    chartbeatConfig,
```
I think we could get rid of the hook dependencies that aren't been used here

andrewscfc replied:
Any idea why they were there in the first place @rhenshaw56 , code been deleted but this not updated?

rhenshaw56 replied:
@andrewscfc I think they were being used [when `getConfig` was called inside the hook](https://github.com/bbc/simorgh/commit/dabc85450d94e01d3ed849d6c7a414213caf6e0c#diff-f85d75438a01e2cb63e2f407f289dfc6) but that's been moved out to line 19

andrewscfc replied:
@rhenshaw56 just found this change causes an infinite render loop!
<img width="673" alt="image" src="https://user-images.githubusercontent.com/9645462/69956509-18e00180-14f8-11ea-94e7-3e462835614b.png">

I'm not quite sure why yet, just looking into it. I may have to back out https://github.com/bbc/simorgh/pull/4861/commits/bd1eb2e871be6a813fc822a6c09790eeb8089610 and put in an eslint exception..

andrewscfc replied:
~~Should be resolved by: 8a47c6aebbc6ec275c042fc0570431aa3b4c6bef~~

andrewscfc replied:
As below, took resolved by e91c6d37ee509da4bc8b039dd5143aea2bb0c3b3

rhenshaw56 reviewed (approved):
Looks good! Tested locally and the error isn't apparent anymore.

oluoluoxenfree committed:
Merge branch 'latest' into chartbeat-amp

andrewscfc committed:
refactor `ChartbeatAnalytics` to be more
maintainable, there only being on list of
hook dependencies. Eslint rule supressed
as is not useful in this case

Co-Authored-By: Jonathan Roebuck <jonathan.roebuck@bbc.co.uk>

andrewscfc committed:
Change eslint approach to use `useMemo`
as suggested by React team:
https://github.com/facebook/react/issues/14476#issuecomment-471199055

jroebu14 commented:
This approach makes sense to me but I'm just conscious that we're using memoization for something that's not performance optimisation. Also, reading from the `useMemo` docs:

> You may rely on useMemo as a performance optimization, not as a semantic guarantee. In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without useMemo — and then add it to optimize performance.

andrewscfc committed:
Revert "Change eslint approach to use `useMemo`"

This reverts commit 8a47c6aebbc6ec275c042fc0570431aa3b4c6bef.

andrewscfc commented:
Reply to @jroebu14 
> This approach makes sense to me but I'm just conscious that we're using memoization for something that's not performance optimisation. Also, reading from the `useMemo` docs:

>> You may rely on useMemo as a performance optimization, not as a semantic guarantee. In the future, React may choose to “forget” some previously memoized values and recalculate them on next render, e.g. to free memory for offscreen components. Write your code so that it still works without useMemo — and then add it to optimize performance.

I think I agree with @jroebu14 here we are using a side-effect of `useMemo` not the API's primary purpose, performance optimisation. I have reverted the memo commit https://github.com/bbc/simorgh/pull/4861/commits/bb1ea3fd0a42d69353cf1df229a0faff954a48a7 and gone back to 
https://github.com/bbc/simorgh/pull/4861/commits/e91c6d37ee509da4bc8b039dd5143aea2bb0c3b3 me and @jroebu14 paired on.

I have updated the description above to explain the apporach.

andrewscfc committed:
Merge branch 'latest' into chartbeat-amp

PriyaKR committed:
Merge branch 'latest' into chartbeat-amp

PriyaKR committed:
Merge branch 'latest' into chartbeat-amp

PriyaKR committed:
Merge branch 'latest' into chartbeat-amp

PriyaKR commented:
Testing:
Noticed the below issues

- I could see this issue happening again https://github.com/bbc/simorgh/issues/3977
- Also when navigating to the inline article page ie say from this article http://localhost:7080/news/articles/c6v11qzyv8po.amp to http://localhost:7080/news/articles/c0g992jmmkko i don't find any `r` parameter being set

Commit:
only send beacon when pageData changes

Commit:
Merge branch 'latest' of github.com:bbc/simorgh into chartbeat-amp

jroebu14 commented:
@PriyaKR @jimjohnsonrollings 

- > I could see this issue happening again #3977

I've fixed this issue. We were watching for changes to the values of a few things like the current route path and the article page data returned from the API. These values changed at different times due to the async nature of fetching page data so we would ping chartbeat when the path changed and then ping again when page data was returned. I've changed it to only ping when page data changes because by then we will have all the values we need.

- > Also when navigating to the inline article page ie say from this article > http://localhost:7080/news/articles/c6v11qzyv8po.amp to > http://localhost:7080/news/articles/c0g992jmmkko i don't find any r parameter being set

I think the `r` value is specific to amp as the referrer. On canonical the referrer is `v` as in virtual referrer because we are on a single page app. So I think this is not a bug and expected behaviour. 

My only question about this is should we send the referrer `r` value on the initial visit of Simorgh (when being referred from outside Simorgh, including AMP) and then send the virtual referrer `v` value on onward journeys within the Simorgh app?

jimjohnsonrollings commented:
@jroebu14 On PAL, it looks like the initial external referrer is sent in the `r` value for the first page, and then the `v` is used on subsequent pages, despite PAL not being an SPA.  I suspect this is because the `v` value is not intended to be the _actual_ URL of the origin page but the _canonical_ URL of the origin page.  So I think it probably would be correct that the `r` field is used for the initial page view referred from externally, although I'm not sure it actually would make much difference which we use.

jroebu14 commented:
@jimjohnsonrollings ok thanks.

So the options are:
- fix this PR so that `r` is sent on initial page view (and not `v`) with `v` on subsequent pages
- leave this PR and create a separate issue to do the above mentioned
- do nothing for now and carry out the work mentioned above if it is identified as a problem

jimjohnsonrollings commented:
@jroebu14 I like Option 3 I think

jroebu14 commented:
@jimjohnsonrollings cool 🙂

PriyaKR commented:
> @PriyaKR
> 
> * > I could see this issue happening again #3977
>   > I've fixed this issue in the last commit. We were watching for changes to the values of a few things like the current route path and the article page data returned from the API. These values changed at different times due to the async nature of fetching page data so we would ping chartbeat when the path changed and then ping again when page data was returned. I've changed it to only ping when page data changes.

_I am still seeing the virtual update happening twice when using the back button_
> * > Also when navigating to the inline article page ie say from this article > http://localhost:7080/news/articles/c6v11qzyv8po.amp to > http://localhost:7080/news/articles/c0g992jmmkko i don't find any r parameter being set
>   > I think the `r` value is specific to amp as the referrer. On canonical the referrer is `v` as in virtual referrer because we are on a single page app. So I think this is not a bug and expected behaviour.
> 
> My only question about this is should we send the referrer `r` value on the initial visit of Simorgh (when being referred from outside Simorgh, including AMP) and then send the virtual referrer `v` value on onward journeys within the Simorgh app?

PriyaKR committed:
Merge branch 'latest' into chartbeat-amp

PriyaKR commented:
Ok looks good to me.The t value is supposed to be unique for a page and not for the session.
Also the chartbeat pings on amp page looks good.
